content/blog/2024/pkg-importer.md

```diff
@@ -0,0 +1,198 @@
+
+This was a fairly simple Vite project, so I hadn't created a configuration file
+at `vite.config.ts` yet. Once I created that, I made my changes to the
+`css.preprocessorOptions.scss` object.
+
+Note that changes in the `scss` object only will affect `.scss` files. If you're
+using the indent-based `.sass` format instead, you'll need to add the Node
+package importer to `css.preprocessorOptions.sass`.
+
+Vite currently only supports the legacy API, so I just needed to set the
+`pkgImporter` option.
+
+```ts
+// ./vite.config.ts
+import * as sass from 'sass';
+import { defineConfig } from 'vite'
+export default defineConfig({
+  css: {
+    preprocessorOptions: {
+      scss: {
+        pkgImporter: new sass.NodePackageImporter()
+      },
+    },
+  },
+});
+```
+
+With that in place, I can now update my Sass import.
+
+Before:
+
+```css
+/* File: ./src/styles/main.scss */
+@use 'vuetify' with (
+```
+
+After:
+
+```css
+@use 'pkg:vuetify' with (
+```
+
+But wait, that's actually more characters- why would I make that change?
+
+Before, the location of `vuetify` was being determined by Vite. Adding `pkg:`
+moves that logic into Sass. This doesn't have an immediate impact in this case,
+but it does make the code more portable. My styles could be shared with someone
+building with Webpack or ported to a future build tool. When more Package
+importers become available, this code even be used outside a Node module.
+
+## Things to watch out for
+
+While this change, in most cases, will be straightforward, you do want to make
+sure that you are still importing the same file. While porting over the
+Bootstrap project in the Webpack example, I also wanted to update the import for
+`bootstrap-icons`.
+
+Before:
+
+```css
+/* File: ./src/styles/index.scss */
+@import '../../node_modules/bootstrap-icons/font/bootstrap-icons.css';
+```
+
+However, when I updated that import to `pkg:bootstrap-icons`, I realized I was
+now loading `bootstrap-icons.scss`, and not the precompiled CSS file. This is
+because `bootstrap-icons` has a `sass` key in the `package.json` with the value
+`"sass": "font/bootstrap-icons.scss"`, so the Sass Node package importer
+identifies that as the desired import.
+
+Functionally, this worked, but meant that I was needing to compile
+`bootstrap-icons` along with my dev styles, slowing things down. The Node
+package importer allows for relative paths within a module, so I tried
+`pkg:bootstrap-icons/font/bootstrap-icons.css`. However, Webpack was resolving
+files ending with `.css` with `css-loader`, which doesn't understand `pkg:`
+URLs.
+
+At this point, I have a few options. I could import the `.scss` file, tweak the
+Webpack config settings for `css-loader` to not try to resolve `pkg:` URLs, or
+leave it as is. Because I knew I had a story in my backlog to address some other
+issues with `bootstrap-icons`, I decided to leave it as is, and address this as
+part of the upcoming story.
```

Comment:
@jgerigmeyer This is a somewhat unsatisfying conclusion, but I wasn't able to find a way to get Webpack to ignore @imports that start with `pkg`- any ideas?

Reply:
Hrm, yeah. I think it's okay to note some of the options here, and I might consider removing the "Because I knew I had..." final sentence.

I haven't tried it, but I'm pretty sure `sass-loader` can parse CSS fine -- would it make sense to configure Webpack to run CSS files through sass-loader first, just like it does with `.scss` files? If so, you could amend this paragraph to:

```
At this point, I have a few options. I could import the `.scss` file and compile
it, tweak the Webpack config settings to use `sass-loader` for `.css` files, or
leave it as is.
```

Reply:
I agree, good call.